### PR TITLE
Validator announce

### PR DESCRIPTION
**Pull request by asaj**

### CODE REVIEW

File: HIP-6.md
Comment (tkporter) [inline HIP-6.md:23]: Do these validator announcements only apply to validators that plan to sign checkpoints in the manner outlined by #6?
Wondering if we should add a "signatureType" or something to allow for other types of attestations in the future (e.g. different signature schemes or proofs or something).
Also wondering if we should explicitly outline the expected content of the S3 buckets provided here somewhere- e.g. document `latest_checkpoint.json` and how each signed checkpoint should be provided

File: HIP-6.md
Comment (tkporter) [inline HIP-6.md:113]: Ive never been a huge fan of relying upon bools for success and have generally preferred reverting upon an error occurring. Wondering if there's a motivation behind bool over reverting?

File: HIP-6.md
Comment (asaj) [reply HIP-6.md:133]: Yes, though the previous entry will only be overwritten if they register with the same storage modaility.

Although actually I think the current proposal doesn't have a way to disambiguate which signature was more recent, so I guess it wouldn't be recommended.

If you did, someone could just re-submit your original announcement and overwrite your new location

File: HIP-6.md
Comment (asaj) [reply HIP-6.md:23]: > Do these validator announcements only apply to validators that plan to sign checkpoints in the manner outlined by https://github.com/hyperlane-xyz/hips/pull/6?

I hadn't really thought about other uses. What were you imagining?

> Also wondering if we should explicitly outline the expected content of the S3 buckets provided here somewhere- e.g. document latest_checkpoint.json and how each signed checkpoint should be provided

Agreed that this could use specification. Maybe in its own AWS S3 storage HIP? Or alternatively in #6 ?

File: HIP-6.md
Comment (asaj) [reply HIP-6.md:113]: No reason other than following a common pattern, I was imagining this function would return true if successful and revert if not

File: HIP-6.md
Comment (nambrot) [inline HIP-6.md:29]: IMO the binary should provide a separate command to create this attestation on demand?

File: HIP-6.md
Comment (nambrot) [inline HIP-6.md:37]: Could a URL based format with the protocol specifying the type be sufficient?

File: HIP-6.md
Comment (nambrot) [inline HIP-6.md:68]: This is more for my own personal curiosity, but what other storage types that are not URLs would you imagine?

File: HIP-6.md
Comment (nambrot) [reply HIP-6.md:133]: It would be nice to be able for Validators to revoke a previous registration, as well as support multiple registrations for the same storage modality (i.e. different regions of AWS)

Comment (asaj) [reply HIP-6.md:133]: Care to make an argument for this as being part of the initial scope? IMO these features can and should be omitted so we can ship faster.

File: HIP-6.md
Comment (asaj) [reply HIP-6.md:29]: Why? Seems much more convenient to have it just show up without you needing to do anything

File: HIP-6.md
Comment (asaj) [reply HIP-6.md:37]: Not necessarily (e.g. local storage). Given that we don't know what the other types may be it doesn't feel prudent to make that sort of assumption.

File: HIP-6.md
Comment (asaj) [reply HIP-6.md:68]: IPFS CID? But tbh I think that misses the point, we shouldn't have to try to guess when it's so easy to make it more general than a URL

File: HIP-6.md
Comment (asaj) [reply HIP-6.md:133]: It's worth noting that these features can easily be added in a backwards compatible way in the future.

File: HIP-6.md
Comment (yorhodes) [reply HIP-6.md:23]: > I hadn't really thought about other uses. What were you imagining?

any protocol could tap into staked hyperlane validators for security on arbitrary data maybe?

Comment (yorhodes) [reply HIP-6.md:23]: >Agreed that this could use specification. Maybe in its own AWS S3 storage HIP? Or alternatively in https://github.com/hyperlane-xyz/hips/pull/6 ?

IMO we can spec this API and schema here, independently of storage provider

File: HIP-6.md
Comment (yorhodes) [reply HIP-6.md:68]: this feels like it fits in another HIP

File: HIP-6.md
Comment (yorhodes) [reply HIP-6.md:37]: somewhat agree that storageType enum here is not only superfluous but potentially confusing to configure outside of solidity

File: HIP-6.md
Comment (yorhodes) [inline HIP-6.md:103]: the return type feels easy to misinterpret
is it important to support multiple modalities for a single validator?

File: HIP-6.md
Comment (asaj) [reply HIP-6.md:37]: What would you propose then? Just bytes with the leading byte as the type?

File: HIP-6.md
Comment (asaj) [reply HIP-6.md:103]: No, IMO, it just seemed very easy to do. If we fold the type into a bytes string, would the return type look better to you?

Comment (yorhodes) [reply HIP-6.md:103]: my comment is more about the fact that the indices of the input do not match the output

Comment (asaj) [reply HIP-6.md:103]: Ah yeah we could do nested arrays instead

File: HIP-6.md
Comment (nambrot) [reply HIP-6.md:29]: I guess a separate commend feels more intuitive for me for a discrete output like this but I don't feel strongly. Having to SIGINT just to get that value seems unusual

File: HIP-6.md
Comment (nambrot) [reply HIP-6.md]: What would not be sufficient about  a URL for local storage? Imo that feels pretty standard (including the protocol at the beginning like file or S3)

File: HIP-6.md
Comment (nambrot) [reply HIP-6.md:123]: I think an example would be to specify multiple S3 buckets for availability reasons. Or if a validator wants to move from S3 to  GCS. I'd imagine supporting this would be trivial but I don't feel strongly

Comment (asaj) [reply HIP-6.md:126]: Updated spec allows for multiple registrations, but still doesn't support de-registration.

File: HIP-6.md
Comment (asaj) [reply HIP-6.md:23]: @yorhodes we *can*, and probably should at some point, but I'm not sure writing up a spec for what's already implemented would be as useful as some of the other things we could be working on

File: HIP-6.md
Comment (asaj) [reply HIP-6.md]: IMO sticking it in the storage modality under `announcement.json` is very convenient. If you're running a validator, it will be there, accessible to you, to register at any time.

File: HIP-6.md
Comment (asaj) [reply HIP-6.md]: I think I've encoded what you're looking for

File: HIP-6.md
Comment (asaj) [reply HIP-6.md]: Made it nested arrays instead

File: HIP-6.md
Comment (nambrot) [reply HIP-6.md]: Is the plan then for the contract to be upgradable? If this contract is not upgradable then I feel decently strong that de-registration should be be supported, even if just at the contract level. It seems very trivial to add such a feature

File: HIP-6.md
Comment (nambrot) [reply HIP-6.md:23]: Agree that specifying this is useful but ultimately a nice to have

File: HIP-6.md
Comment (nambrot) [inline HIP-6.md:53]: Like I mentioned, I would personally advocate for supporting addition and removals in the contract right now to avoid the need to change it later, so here would probably add "add" or "remove" to the digest, but not going to block on this

**Review by nambrot — APPROVED** [review]
Approving with disagree but commit on the deregistration of announcements

**Pull request merged**
